core/modules/gtpu_encap.cc

```diff
@@ -281,83 +126,38 @@ void GtpuEncap::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+    iph->src = (be32_t)(at_tout_sip);
+    iph->dst = (be32_t)(at_tout_dip);
     EmitPacket(ctx, p, FORWARD_GATE);
-    /* increment hit idx */
-    j++;
   }
 }
 /*----------------------------------------------------------------------------------*/
-void GtpuEncap::DeInit() {
-  uint32_t next = 0;
-  uint64_t *key;
-  void *_data;
-  int rc;
-  do {
-    rc = rte_hash_iterate(session_map, (const void **)&key, &_data, &next);
-    if (rc >= 0) {
-      struct session_info *data = (struct session_info *)_data;
-      /* now remove the record */
-      if (rte_hash_del_key(session_map, key) < 0) {
-        uint32_t ip = UE_ADDR(*key);
-        std::cerr << "Failed to remove record with UE address: "
-                  << ToIpv4Address(be32_t(ip)) << std::endl;
-      }
-      rte_free(data);
-      /* resetting back to NULL */
-      next = 0;
-    }
-  } while (rc >= 0);
-
-  /* finally free the hash table */
-  rte_hash_free(session_map);
-  session_map = NULL;
-}
-/*----------------------------------------------------------------------------------*/
-CommandResponse GtpuEncap::Init(const bess::pb::GtpuEncapArg &arg) {
-  s1u_sgw_ip = arg.s1u_sgw_ip();
-
-  if (s1u_sgw_ip == 0)
-    return CommandFailure(EINVAL, "Invalid S1U SGW IP address!");
-
-  InitNumSubs = arg.num_subscribers();
-  if (InitNumSubs == 0)
-    return CommandFailure(EINVAL, "Invalid number of subscribers!");
-
-  std::string hashtable_name = "session_map" + this->name();
-  std::cerr << "Creating rte_hash: " << hashtable_name << std::endl;
-
-  struct rte_hash_parameters session_map_params = {
-      .name = hashtable_name.c_str(),
-      .entries = (unsigned int)InitNumSubs,
-      .reserved = 0,
-      .key_len = sizeof(uint64_t),
-      .hash_func = rte_jhash,
-      .hash_func_init_val = 0,
-      .socket_id = (int)rte_socket_id(),
-      .extra_flag = RTE_HASH_EXTRA_FLAGS_RW_CONCURRENCY};
-
-  session_map = rte_hash_create(&session_map_params);
-  if (session_map == NULL)
-    return CommandFailure(ENOMEM, "Unable to create rte_hash table: %s\n",
-                          "session_map");
+CommandResponse GtpuEncap::Init(const bess::pb::EmptyArg &) {
+  using AccessMode = bess::metadata::Attribute::AccessMode;
+  tout_sip_attr = AddMetadataAttr("tunnel_out_src_ip4addr", sizeof(uint32_t),
+                                  AccessMode::kRead);
+  DLOG(INFO) << "tout_sip_attr: " << tout_sip_attr << std::endl;
+  tout_dip_attr = AddMetadataAttr("tunnel_out_dst_ip4addr", sizeof(uint32_t),
+                                  AccessMode::kRead);
+  DLOG(INFO) << "tout_dip_attr: " << tout_dip_attr << std::endl;
+  tout_teid =
+      AddMetadataAttr("tunnel_out_teid", sizeof(uint32_t), AccessMode::kRead);
+  DLOG(INFO) << "tout_teid: " << tout_teid << std::endl;
+  tout_uport = AddMetadataAttr("tunnel_out_udp_port", sizeof(uint16_t),
+                               AccessMode::kRead);
+  DLOG(INFO) << "tout_uport: " << tout_uport << std::endl;
```

Comment:
replace vars with ret and check if successful on each one and keep the DLOG. return commandresponse grpc error on failure

Reply:
See comment above.